analysis/pseudobulk-bulk-prediction/exploratory-notebooks/permute-gsea.Rmd

```diff
@@ -0,0 +1,193 @@
+      dplyr::group_by(ensembl_id) |>
+      dplyr::summarize(median_res = median(residuals)) |>
+      dplyr::arrange(-median_res) |>
+      dplyr::pull(median_res, name = ensembl_id)
+  }
+)
+```
+
+## Prepare shuffled gene sets
+
+Define the gene set:
+
+```{r}
+msigdbr::msigdbr_collections() |>
+  dplyr::filter(gs_cat == params$msigdbr_category)
+geneset_df <- msigdbr::msigdbr(species = "Homo sapiens", category = params$msigdbr_category)
+```
+
+Note that it contains `r nrow(geneset_df)` gene sets.
+
+
+```{r}
+geneset_count_df <- geneset_df |>
+  dplyr::count(gs_name)
+
+unique_ensembl_ids <- geneset_df |>
+  dplyr::distinct(human_ensembl_gene) |>
+  dplyr::pull(human_ensembl_gene)
+```
+
+
+We'll create `r params$reps` shuffled versions of the gene sets.
+
+```{r}
+# Create a shuffled version of genesets
+generate_shuffle <- function(geneset_count_df, ensembl_ids){
+  geneset_count_df |>
+    purrr::pmap(
+    \(gs_name, n) {
+      data.frame(
+        gs_name = gs_name, 
+        ensembl_id = sample(ensembl_ids, n) # no replacement!
+      )
+  }) |>
+  purrr::list_rbind()
+}
+
+shuffled_geneset_df <- 1:params$reps |>
+  purrr::map(
+    \(i) {generate_shuffle(geneset_count_df, unique_ensembl_ids)}
+  )
+
+# Ensure correct lengths
+purrr::walk(
+  shuffled_geneset_df, 
+  \(df) {stopifnot(nrow(df) == nrow(geneset_df))}
+)
+```
+
+
+## Perform GSEA 
+
+Next, we'll run GSEA on them all:
+
+(This next chunk takes a little bit to run; we'll turn off messages and warnings since this is quite a lot of GSEAs!)
+
+```{r, message = FALSE, warning = FALSE}
+shuffled_gsea_df <- ranked_residuals_list |>
+  purrr::map(
+  \(ranked_genes) {
+    
+    purrr::map(
+      shuffled_geneset_df, 
+      \(shuffled_rep) {
+        gsea_result <- clusterProfiler::GSEA(
+          geneList = ranked_genes, 
+          TERM2GENE = shuffled_rep
+        )
+        gsea_result@result |>
+          dplyr::summarize(
+            bulk_enriched = sum(NES > 0), 
+            pseudobulk_enriched = sum(NES < 0)
+          )
+      }
+    ) |> purrr::list_rbind(names_to = "rep")
+    
+  }) |> 
+  purrr::list_rbind(names_to = "project_id") |>
+  # format data for plotting
+  tidyr::pivot_longer(
+    contains("enriched"), 
+    names_to = "direction", 
+    values_to = "count"
+  ) |>
+  # pseudobulk should be first
+  dplyr::mutate(direction = forcats::fct_rev(direction))
+```
+
+## Visualize number of significant pathways
+
+Finally, we'll make a histogram of how many signficant gene sets were identified across the `r params$reps` replicates.
+Note that this data considers all results GSEA deemed to be signficant; there is no further filtering by effect size or P-value at this time.
+
+```{r, fig.width = 10, fig.height = 3, message = FALSE}
+ggplot(shuffled_gsea_df) + 
+  aes(x = count, fill = direction) + 
+  geom_histogram() + 
+  facet_wrap(vars(project_id), nrow = 1) +
+  scale_fill_brewer(palette = "Dark2") +
+  theme(legend.position = "bottom")
+```
```

Comment:
Returning this while we ponder adding in real results. Can you give this a nicer x-axis label? Both axes are labeled with "count," and I found it confusing.